tests/copy_test.go

```diff
@@ -0,0 +1,278 @@
+		case tar.TypeDir:
+			dirs = append(dirs, f)
+			if err := os.MkdirAll(p, f.mode); err != nil {
+				return err
+			}
+			dir, err := os.Open(p)
+			if err != nil {
+				return err
+			}
+			if err := dir.Chmod(f.mode); err != nil {
+				dir.Close()
+				return err
+			}
+			dir.Close()
+		case tar.TypeReg:
+			err := ioutil.WriteFile(p, []byte(f.contents), f.mode)
+			if err != nil {
+				return err
+			}
+
+		}
+
+		if err := os.Chtimes(p, f.atime, f.mtime); err != nil {
+			return err
+		}
+	}
+
+	// Restore dirs atime and mtime. This has to be done after extracting
+	// as a file extraction will change its parent directory's times.
+	for _, d := range dirs {
+		p := filepath.Join(dir, d.path)
+		if err := os.Chtimes(p, d.atime, d.mtime); err != nil {
+			return err
+		}
+	}
+
+	defaultTime := time.Unix(0, 0)
+	// Restore the base dir time as it will be changed by the previous extractions
+	if err := os.Chtimes(dir, defaultTime, defaultTime); err != nil {
+		return err
+	}
+	return nil
+}
+
+func TestCopyOneDir(t *testing.T) {
+	workingDir := setUpTest(t)
+	defer cleanUpTest(workingDir)
+
+	sourceDir, err := ioutil.TempDir("", "acbuild-test-copy")
+	if err != nil {
+		panic(err)
+	}
+	defer os.RemoveAll(sourceDir)
+
+	time1 := time.Now()
+	//time2 := time.Now().Add(1 * time.Microsecond)
+	files := []*buildFileInfo{
+		&buildFileInfo{path: "file01", typeflag: tar.TypeReg, mode: 0644, atime: time1, mtime: time1, contents: "hello"},
+		&buildFileInfo{path: "dir01", typeflag: tar.TypeDir, mode: 0755, atime: time1, mtime: time1},
+		&buildFileInfo{path: "dir01/file01", typeflag: tar.TypeReg, mode: 0644, atime: time1, mtime: time1, contents: "hello"},
+	}
+
+	err = buildFS(sourceDir, files)
+	if err != nil {
+		panic(err)
+	}
+
+	err1 := runACBuild(workingDir, "copy", sourceDir, dest)
+	if err1 != nil {
+		t.Fatalf("%v\n", err1)
+	}
+
+	testMatchingFSTree(t, workingDir, sourceDir, dest)
+}
+
+func TestCopyOneFile(t *testing.T) {
+	workingDir := setUpTest(t)
+	defer cleanUpTest(workingDir)
+
+	sourceFile, err := ioutil.TempFile("", "acbuild-test-copy")
+	if err != nil {
+		panic(err)
+	}
+	_, err = sourceFile.Write([]byte("this is a test file"))
+	if err != nil {
+		panic(err)
+	}
+	sourceFile.Close()
+	defer os.RemoveAll(sourceFile.Name())
+
+	err1 := runACBuild(workingDir, "copy", sourceFile.Name(), dest)
+	if err1 != nil {
+		t.Fatalf("%v\n", err1)
+	}
+
+	testMatchingFSTree(t, workingDir, sourceFile.Name(), dest)
+}
+
+func TestCopyOneDirToExistingDir(t *testing.T) {
+	workingDir := setUpTest(t)
+	defer cleanUpTest(workingDir)
+
+	sourceDir, err := ioutil.TempDir("", "acbuild-test-copy")
+	if err != nil {
+		panic(err)
+	}
+	defer os.RemoveAll(sourceDir)
+
+	err = os.MkdirAll(path.Join(workingDir, ".acbuild", "currentaci", aci.RootfsDir, dest), 0755)
+	if err != nil {
+		panic(err)
+	}
+
+	err1 := runACBuild(workingDir, "copy", sourceDir, dest)
+	if err1 == nil {
+		t.Fatalf("got no error when destination directory exists, was expecting one")
```

Comment:
Also, if possible this kind of test should check for the actual error type to avoid the risk of false positives